Add `alloc::rc::UniqueRc`

eholk opened the pull request:
This PR implements `UniqueRc` as described in https://github.com/rust-lang/libs-team/issues/90.

I've tried to stick to the API proposed there, incorporating the feedback from the ACP review. For now I've just implemented `UniqueRc`, but we'll want `UniqueArc` as well. I wanted to get feedback on this implementation first since the `UniqueArc` version should be mostly a copy/paste/rename job.

eholk committed:
Introduce `alloc::::UniqueRc`

This is an `Rc` that is guaranteed to only have one strong reference.
Because it is uniquely owned, it can safely implement `DerefMut`, which
allows programs to have an initialization phase where structures inside
the `Rc` can be mutated.

The `UniqueRc` can then be converted to a regular `Rc`, allowing sharing
and but read-only access.

During the "initialization phase," weak references can be created, but
attempting to upgrade these will fail until the `UniqueRc` has been
converted to a regular `Rc`. This feature can be useful to create
cyclic data structures.

This API is an implementation based on the feedback provided to the ACP
at https://github.com/rust-lang/libs-team/issues/90.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5050212585/jobs/9060634783) [(plain)](https://github.com/rust-lang/rust/commit/ef6d1fa6b550aaa7e4c6c86a176c05290a75bd7f/checks/13671855984/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
............................................

failures:

---- rc::tests::test_unique_rc_drop_weak stdout ----
error: test failed, to rerun pass `-p alloc --lib`
thread 'rc::tests::test_unique_rc_drop_weak' panicked at 'attempt to subtract with overflow', library/alloc/src/rc.rs:2683:29

failures:
failures:
    rc::tests::test_unique_rc_drop_weak
test result: FAILED. 395 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 458.12ms

Build completed unsuccessfully in 0:01:07
```

</details>

eholk commented:
@rustbot author

There are a few issues that need fixed before this is ready for review.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5051034458/jobs/9062438663) [(plain)](https://github.com/rust-lang/rust/commit/ee8c5addc196707addf46a4303c39f1beca9673c/checks/13674217192/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 finished in 58.964 seconds
Check compiletest suite=run-make mode=run-make (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 324 tests
..............i....ii....ii..F..........................................................  88/324
.........iiiiiii.......i.ii.iiiii...................................iii................. 264/324
...............ii.i......i...iiiiiii.iii.ii.i...............

failures:
failures:
Build completed unsuccessfully in 0:33:46

---- [run-make] tests/run-make/alloc-no-oom-handling stdout ----

error: make failed
status: exit status: 2
command: cd "/checkout/tests/run-make/alloc-no-oom-handling" && AR="ar" CC="cc -ffunction-sections -fdata-sections -fPIC -m64" CXX="c++ -ffunction-sections -fdata-sections -fPIC -m64" HOST_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" LD_LIB_PATH_ENVVAR="LD_LIBRARY_PATH" LLVM_BIN_DIR="/usr/lib/llvm-14/bin" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgputargetmca amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwp engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils ve veasmparser vecodegen vectorize vedesc vedisassembler veinfo webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" LLVM_FILECHECK="/usr/lib/llvm-14/bin/FileCheck" NODE="/usr/bin/node" PYTHON="/usr/bin/python3" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTDOC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" RUST_BUILD_STAGE="stage2-x86_64-unknown-linux-gnu" RUST_DEMANGLER="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin/rust-demangler" S="/checkout" TARGET="x86_64-unknown-linux-gnu" TARGET_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" TMPDIR="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/alloc-no-oom-handling/alloc-no-oom-handling" "make"
--- stdout -------------------------------
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/alloc-no-oom-handling/alloc-no-oom-handling:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/alloc-no-oom-handling/alloc-no-oom-handling -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/alloc-no-oom-handling/alloc-no-oom-handling  --edition=2021 -Dwarnings --crate-type=rlib ../../../library/alloc/src/lib.rs --cfg no_global_oom_handling
--- stderr -------------------------------
--- stderr -------------------------------
error[E0599]: no function or associated item named `new` found for struct `Box<_, _>` in the current scope
    --> ../../../library/alloc/src/rc.rs:2803:33
     |
2803 |               ptr: Box::leak(Box::new(RcBox {
     |                                   ^^^ function or associated item not found in `Box<_, _>`
    ::: ../../../library/alloc/src/boxed.rs:195:1
     |
195  | / pub struct Box<
195  | / pub struct Box<
196  | |     T: ?Sized,
197  | |     #[unstable(feature = "allocator_api", issue = "32838")] A: Allocator = Global,
198  | | >(Unique<T>, A);
     | |_- function or associated item `new` not found for this struct
error: aborting due to previous error

For more information about this error, try `rustc --explain E0599`.
make: *** [Makefile:4: all] Error 1
```

</details>

programmerjake commented:
to me, something named `UniqueRc<T>` should contain a `T` and implement `DerefMut<Target = T>` and basically be `Rc` but with the restriction that `strong_count == 1` and `Weak::upgrade` doesn't yet work...

what you have here seems more like a `UninitalizedRc` or `EmptyRc` where it allocated space for a `RcBox` but didn't fill it in yet.

m-ou-se commented:
> to me, something named `UniqueRc<T>` should contain a `T` and implement `DerefMut<Target = T>` and basically be `Rc` but with the restriction that `strong_count == 1` and `Weak::upgrade` doesn't yet work...

That's indeed what I meant in https://github.com/rust-lang/libs-team/issues/90#issuecomment-1553201022

eholk commented:
Ah, I see what you mean. I misinterpreted the original suggestion.

`UniqueRc` didn't seem like the right name for what I've implemented. While I was writing the documentation, something like `PlaceholderRc<T>` felt like a better name.

So with `UniqueRc`, creating a cyclic structure would look more like this, right?

```rust
struct Cyclic {
    this: Weak<Cyclic>,
}

let mut cycle = UniqueRc::new(Cyclic { this: Weak::new() };
cycle.this = UniqueRc::weak(cycle);

// convert to a normal `Rc`
let rc = UniqueRc::into_rc(cycle);
```

I'm curious why the Libs API team prefers this approach? I personally find something like `PlaceholderRc` or `UninitializedRc` more ergonomic, but I admit this is a matter of preference and I am likely overlooking other concerns around consistency with the rest of std, etc.

One nice property I see for `PlaceholderRc` is that it's pretty straightforward to implement `Rc::new_cyclic` in terms of `PlaceholderRc`. It's less obvious how to do this with `UniqueRc`, although @joboet's trick of using `UniqueRc<MaybeUninit<T>>` should work.

I appreciate the discussion here! One thing this has pointed out to me is that it's actually possible to implement this without depending on any internal details of `Rc`, so it could be done in an external crate (although it would require unsafe code, of course). In our usage of this, we had basically written our own `RcBox` and transmuted the contents of the `Rc` into that, which worked but was incredibly fragile. Given the realization we can do this without depending on internal details, I wonder if it makes more sense to try this as an external crate first?

programmerjake commented:
for `UniqueRc` at least, it does require modifying `Rc`'s internals in order to prevent `Weak::upgrade` from working until the `UniqueRc` has been converted into an `Rc`, otherwise you could have both shared and mutable references to `T` which is UB.

eholk commented:
> for `UniqueRc` at least, it does require modifying `Rc`'s internals in order to prevent `Weak::upgrade` from working until the `UniqueRc` has been converted into an `Rc`

Oh, good point. Thanks for pointing that out!

eholk commented:
I updated the PR to implement `UniqueRc<T>` as a thing that always holds an initialized `T` and provides a `DerefMut` impl so you can freely mutate its contents before converting it to an `Rc`.

I also changed all the functions to not have a self, since the pattern for `Rc` and such is to avoid adding new methods that could conflict with methods on the underlying `T`.

Finally, I renamed `UniqueRc::weak` to `UniqueRc::downgrade` for better symmetry with `Rc`.

marmeladema commented on library/alloc/src/rc.rs:
```diff
@@ -2746,3 +2746,128 @@ fn data_offset_align(align: usize) -> usize {
+impl<T> UniqueRc<T> {
+    /// Creates a new `UniqueRc`
+    ///
+    /// Weak references to this `UniqueRc` can be created with [`UniqueRc::weak`]. Upgrading these
```
`downgrade` instead of `weak`?

eholk replied:
Good catch! Fixed.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5074033001/jobs/9113770700) [(plain)](https://github.com/rust-lang/rust/commit/bfe046e726dabcbb06d51ca761ccd6fe09f8b4e4/checks/13740015401/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking core v0.0.0 (/checkout/library/core)
   Compiling compiler_builtins v0.1.91
    Checking rustc-std-workspace-core v1.99.0 (/checkout/library/rustc-std-workspace-core)
 Documenting alloc v0.0.0 (/checkout/library/alloc)
error: unresolved link to `UniqueRc::weak`
     |
     |
2795 |     /// Weak references to this `UniqueRc` can be created with [`UniqueRc::weak`]. Upgrading these
     |                                                                  ^^^^^^^^^^^^^^ the struct `UniqueRc` has no field or associated item named `weak`
     |
     = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
error: could not document `alloc`

Caused by:
Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustdoc --edition=2021 --crate-type lib --crate-name alloc library/alloc/src/lib.rs --target x86_64-unknown-linux-gnu -o /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/doc --cfg 'feature="compiler-builtins-c"' -Zunstable-options --check-cfg 'names()' --check-cfg 'values()' --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat -Z unstable-options --resource-suffix 1.71.0 --markdown-css rust.css --markdown-no-toc --index-page /checkout/src/doc/index.md -C metadata=3ede8d0c18b54098 -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/release/deps --extern compiler_builtins=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libcompiler_builtins-82f50f274c4bfe7f.rmeta --extern core=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libcore-77221b083c66e738.rmeta --cfg=bootstrap -Csymbol-mangling-version=legacy -Zunstable-options -Zunstable-options '--check-cfg=values(bootstrap)' '--check-cfg=values(stdarch_intel_sde)' '--check-cfg=values(no_fp_fmt_parse)' '--check-cfg=values(no_global_oom_handling)' '--check-cfg=values(no_rc)' '--check-cfg=values(no_sync)' '--check-cfg=values(freebsd12)' '--check-cfg=values(freebsd13)' '--check-cfg=values(backtrace_in_libstd)' '--check-cfg=values(target_env,"libnx")' '--check-cfg=values(target_arch,"asmjs","spirv","nvptx","xtensa")' --document-private-items --document-hidden-items -Dwarnings '-Wrustdoc::invalid_codeblock_attributes' --crate-version '1.71.0-nightly
  (681df1787
  2023-05-24)' '-Zcrate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/")' '-Zcrate-attr=warn(rust_2018_idioms)'` (exit status: 1)
```

</details>

programmerjake commented on library/alloc/src/rc.rs:
```diff
@@ -2746,3 +2746,128 @@ fn data_offset_align(align: usize) -> usize {
+        // SAFETY: This pointer was allocated at creation time so we know it is valid. We know we
+        // have unique ownership and therefore it's safe to make a mutable reference because
+        // `UniqueRc` owns the only strong reference to itself.
+        unsafe { &mut self.ptr.as_mut().value }
```
```suggestion
        unsafe { &mut (*self.ptr.as_ptr()).value }
```
you can't create a `&mut RcBox` since other `Weak` still have valid pointers to that `RcBox` and the `&mut RcBox` disallows any other access while the returned `&mut T` is alive, including to the strong and weak counts.

programmerjake commented on library/alloc/src/rc.rs:
```diff
@@ -2746,3 +2746,128 @@ fn data_offset_align(align: usize) -> usize {
+    fn drop(&mut self) {
+        // We want to make sure the pretend weak reference we're holding gets dropped, so convert
+        // ourselves into an actual Weak reference and let its destructor take over.
+        Weak { ptr: self.ptr };
```
you forgot to drop `T`

eholk replied:
🤦‍♂️ Oops. Back when this was `MaybeUninit` I was going out of my way not to drop it and I forgot I needed to change it back. This should be fixed now.

programmerjake commented on library/alloc/src/rc.rs:
```diff
@@ -2746,3 +2746,128 @@ fn data_offset_align(align: usize) -> usize {
+#[unstable(feature = "unique_rc_arc", issue = "none")]
+#[derive(Debug)]
+pub struct UniqueRc<T> {
+    ptr: NonNull<RcBox<T>>,
```
```suggestion
    ptr: NonNull<RcBox<T>>,
    phantom: PhantomData<RcBox<T>>,
```

it should have a `PhantomData` and `Drop` should have `T` marked `#[may_dangle]`, see `Drop for Rc<T>` for details

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5082920177/jobs/9133245267) [(plain)](https://github.com/rust-lang/rust/commit/43072c3f0b25e5e3b8f01c92477ea56032482bf9/checks/13764825530/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling alloc v0.0.0 (/checkout/library/alloc)
   Compiling cfg-if v1.0.0
   Compiling adler v1.0.2
   Compiling rustc-demangle v0.1.21
error[E0063]: missing field `phantom` in initializer of `UniqueRc<T>`
     |
2802 |         Self {
     |         ^^^^ missing `phantom`

```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5083122342/jobs/9133728879) [(plain)](https://github.com/rust-lang/rust/commit/6b795e424123adaec4cd7338e75304e2bd7bda05/checks/13765466358/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 finished in 58.389 seconds
Check compiletest suite=run-make mode=run-make (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 324 tests
..............i....ii....ii...F.........................................................  88/324
.........iiiiiii.......i.ii.iiiii...................................iii................. 264/324
...............ii.i......i...iiiiiii.iii.ii.i...............

failures:
failures:

---- [run-make] tests/run-make/alloc-no-oom-handling stdout ----

error: make failed
status: exit status: 2
command: cd "/checkout/tests/run-make/alloc-no-oom-handling" && AR="ar" CC="cc -ffunction-sections -fdata-sections -fPIC -m64" CXX="c++ -ffunction-sections -fdata-sections -fPIC -m64" HOST_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" LD_LIB_PATH_ENVVAR="LD_LIBRARY_PATH" LLVM_BIN_DIR="/usr/lib/llvm-14/bin" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgputargetmca amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwp engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils ve veasmparser vecodegen vectorize vedesc vedisassembler veinfo webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" LLVM_FILECHECK="/usr/lib/llvm-14/bin/FileCheck" NODE="/usr/bin/node" PYTHON="/usr/bin/python3" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTDOC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" RUST_BUILD_STAGE="stage2-x86_64-unknown-linux-gnu" RUST_DEMANGLER="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools-bin/rust-demangler" S="/checkout" TARGET="x86_64-unknown-linux-gnu" TARGET_RPATH_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" TMPDIR="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/alloc-no-oom-handling/alloc-no-oom-handling" "make"
--- stdout -------------------------------
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/alloc-no-oom-handling/alloc-no-oom-handling:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/alloc-no-oom-handling/alloc-no-oom-handling -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/alloc-no-oom-handling/alloc-no-oom-handling  --edition=2021 -Dwarnings --crate-type=rlib ../../../library/alloc/src/lib.rs --cfg no_global_oom_handling
--- stderr -------------------------------
--- stderr -------------------------------
error[E0599]: no function or associated item named `new` found for struct `Box<_, _>` in the current scope
    --> ../../../library/alloc/src/rc.rs:2803:33
     |
     |
2803 |               ptr: Box::leak(Box::new(RcBox {
     |                                   ^^^ function or associated item not found in `Box<_, _>`
    ::: ../../../library/alloc/src/boxed.rs:195:1
     |
195  | / pub struct Box<
195  | / pub struct Box<
196  | |     T: ?Sized,
197  | |     #[unstable(feature = "allocator_api", issue = "32838")] A: Allocator = Global,
198  | | >(Unique<T>, A);
     | |_- function or associated item `new` not found for this struct
error: aborting due to previous error

For more information about this error, try `rustc --explain E0599`.
make: *** [Makefile:4: all] Error 1
```

</details>

safinaskar commented:
> I updated the PR to implement `UniqueRc<T>` as a thing that always holds an initialized `T`

I like your previous approach more. It is simpler to use

programmerjake commented:
> > I updated the PR to implement `UniqueRc<T>` as a thing that always holds an initialized `T`
> 
> I like your previous approach more. It is simpler to use

imho both APIs are useful, both the `EmptyRc` (like in `Rc::new_cyclic`) and `UniqueRc` that implements `DerefMut<Target = T>` -- they have different use-cases, so for each use-case one and/or the other or neither is best suited for that use-case.

eholk commented:
> > 
> 
> imho both APIs are useful, both the `EmptyRc` (like in `Rc::new_cyclic`) and `UniqueRc` that implements `DerefMut<Target = T>` -- they have different use-cases, so for each use-case one and/or the other or neither is best suited for that use-case.

`EmptyRc` seems like a good name for it. The nice thing is the two modes are pretty similar (you can implement `EmptyRc` as `UniqueRc<MaybeUninit<T>>`, so it doesn't seem like a big deal to include both.

eholk commented:
I think this is ready now. It's tempting to add `EmptyRc` too, but I think it's better to save that for a later PR. This PR focuses on implementing `UniqueRc` as was suggested by the libs API team.

@rustbot ready

Mark-Simulacrum commented on library/alloc/src/rc.rs:
```diff
@@ -2746,3 +2746,139 @@ fn data_offset_align(align: usize) -> usize {
+/// including fallible or async constructors.
+#[unstable(feature = "unique_rc_arc", issue = "none")]
+#[derive(Debug)]
+pub struct UniqueRc<T> {
```
I'm a little worried that this name makes functions like `is_unique` in the old Rc code harder to understand and may lead to bugs -- since the two unique concepts are different.

I wonder if we can expose most of the niceness (except for Deref/Mut) on Weak, checking if the underlying strong count is zero on get()/get_mut() and adding a constructor for it similar to UniqueRcs. But not sure if that is strictly better.

eholk replied:
Thanks for the feedback! I added that as an unresolved question in the tracking issue (#112566).

I do think it'd be good to iterate some on the naming and API a bit. There are definitely some tradeoffs and I think getting some experience using them will help find the best way to do it.

Mark-Simulacrum reviewed (commented):
Implementation looks fine to me. I think the ACP looks like it was positively received, so I'm inclined to say we should file a tracking issue and I can approve this PR -- the naming of the type feels unfortunate as mentioned in my comment, but I wouldn't block this over that.

eholk commented:
I opened #112566 as the tracking issue for this feature.

eholk commented:
@rustbot ready

riking commented:
Please check API consistency with the Rust-for-Linux UniqueArc https://github.com/Rust-for-Linux/linux/blob/rust/rust/kernel/sync/arc.rs#L502-L578

programmerjake commented:
> Please check API consistency with the Rust-for-Linux UniqueArc https://github.com/Rust-for-Linux/linux/blob/rust/rust/kernel/sync/arc.rs#L502-L578

imho API consistency with `Rc<T>` should take priority over Linux's `UniqueArc` where there's a conflict.

Mark-Simulacrum commented:
@rustbot author

r=me with commits squashed

Mark-Simulacrum commented:
I think we can revise API as needed in future PRs since this is unstable (e.g., for renaming or for kernel API consistencies).

eholk commented:
> Please check API consistency with the Rust-for-Linux UniqueArc https://github.com/Rust-for-Linux/linux/blob/rust/rust/kernel/sync/arc.rs#L502-L578

I added this under open questions in the tracking issue (#112566). Thanks for pointing this out to me, it's good to know of another use case for something like this.

eholk commented:
@bors r=@Mark-Simulacrum

bors commented:
:pushpin: Commit 53003cdd86592f8a9844eef5060699e981dacf07 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 53003cdd86592f8a9844eef5060699e981dacf07 -->
<!-- homu: {"type":"Approved","sha":"53003cdd86592f8a9844eef5060699e981dacf07","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 53003cdd86592f8a9844eef5060699e981dacf07 with merge 14803bda0efc94b6c9a90048d868104c1ddf0685...
<!-- homu: {"type":"BuildStarted","head_sha":"53003cdd86592f8a9844eef5060699e981dacf07","merge_sha":"14803bda0efc94b6c9a90048d868104c1ddf0685"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5317115655/jobs/9629154103)
Approved by: Mark-Simulacrum
Pushing 14803bda0efc94b6c9a90048d868104c1ddf0685 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5317115655/jobs/9629154103"},"merge_sha":"14803bda0efc94b6c9a90048d868104c1ddf0685"} -->

rust-timer commented:
Finished benchmarking commit (14803bda0efc94b6c9a90048d868104c1ddf0685): [comparison URL](https://perf.rust-lang.org/compare.html?start=fe7454bf439c93cbe9ac8a8f7fcfacd5a40244c2&end=14803bda0efc94b6c9a90048d868104c1ddf0685&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=fe7454bf439c93cbe9ac8a8f7fcfacd5a40244c2&end=14803bda0efc94b6c9a90048d868104c1ddf0685&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=fe7454bf439c93cbe9ac8a8f7fcfacd5a40244c2&end=14803bda0efc94b6c9a90048d868104c1ddf0685&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.9% | [-3.4%, -1.0%] | 10    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.9% | [-3.4%, -1.0%] | 10    |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=fe7454bf439c93cbe9ac8a8f7fcfacd5a40244c2&end=14803bda0efc94b6c9a90048d868104c1ddf0685&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=fe7454bf439c93cbe9ac8a8f7fcfacd5a40244c2&end=14803bda0efc94b6c9a90048d868104c1ddf0685&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 657.668s -> 655.501s (-0.33%)

RalfJung commented on library/alloc/src/rc.rs:
```diff
@@ -2744,3 +2744,139 @@ fn data_offset_align(align: usize) -> usize {
+///         me: Weak::new(),
+///     });
+///     rc.me = UniqueRc::downgrade(&rc);
+///     Some(UniqueRc::into_rc(rc))
```
Why does `create_gadget` return an `Option`?

eholk replied:
The example evolved a bit from my first version of it. Originally I wanted to show that you could use this to initialize a circular data structure while the initialization phase is fallible, which is something that the existing `new_cyclic` function doesn't handle well. Unfortunately, there's no fallibility left in this example so the `Option` is probably more of a distraction.
